sdk/communication/azure-communication-chat/src/main/java/com/azure/android/communication/chat/implementation/notifications/fcm/PushNotificationClient.java

```diff
@@ -284,4 +310,42 @@ public void run() {
         this.registrationRenewScheduleTimer = new Timer("PushNotificationRegistrarRenewTimer");
         this.registrationRenewScheduleTimer.schedule(task, delayMs);
     }
+
+    private void refreshEncryptionKeys() throws Throwable {
+        if (this.keyGenerator == null) {
+            this.keyGenerator = KeyGenerator.getInstance("AES");
+            this.keyGenerator.init(KEY_SIZE);
+        }
+
+        this.previousCryptoKey = this.cryptoKey;
+        this.previousAuthKey = this.authKey;
+
+        this.cryptoKey = this.keyGenerator.generateKey();
+        this.authKey = this.keyGenerator.generateKey();
+    }
+
+    private String decryptPayload(String encryptedPayload) throws Throwable {
+        this.logger.info(" Decrypting payload.");
+
+        byte[] encryptedBytes = Base64Util.decodeString(encryptedPayload);
+        byte[] f = NotificationUtils.extractEncryptionKey(encryptedBytes);
+        byte[] iv = NotificationUtils.extractInitializationVector(encryptedBytes);
+        byte[] cipherText = NotificationUtils.extractCipherText(encryptedBytes);
+        byte[] hmac = NotificationUtils.extractHmac(encryptedBytes);
+
+        if (NotificationUtils.verifyEncryptedPayload(f, iv, cipherText, hmac, this.authKey)) {
+            return NotificationUtils.decryptPushNotificationPayload(iv, cipherText, this.cryptoKey);
+        }
+
+        // When client has registered a new key, because of eventual consistency, latencies and concurrency,
+        // the old key can still be used by server side for some notifications
+        // Try old key if the new key failed to decrypt the payload.
+        if (NotificationUtils.verifyEncryptedPayload(f, iv, cipherText, hmac, this.previousAuthKey)) {
+            return NotificationUtils.decryptPushNotificationPayload(iv, cipherText, this.previousCryptoKey);
+        }
```

Comment:
Is it possible to use an old key that has expired a long time ago? How often are we expecting users to call the function to renew?

Reply:
For renew, I am following this Recommended rotation process https://skype.visualstudio.com/DefaultCollection/SCC/_wiki/wikis/SCC.wiki/2253/Cryptography?anchor=recommended-rotation-process

Currently, key rotation happens when registration is renewed, which is 90000 seconds (25 hours). In the doc, they also mentioned I should "Start a short countdown for a grace period". I'll update to this.